src/core/components/pin.js

```diff
@@ -0,0 +1,589 @@
+          if (callback.fired) { return }
+          if (err) {
+            callback.fired = true
+            return callback(err)
+          }
+          nodes.forEach((node) => {
+            const mh = node.multihash
+            if (!directPins.has(mh) && !recursivePins.has(mh)) {
+              // not already pinned recursively or directly
+              indirectKeys.add(mh)
+            }
+          })
+          if (doneCount === rKeys.length - 1) {
+            return callback(null, indirectKeys.toStringArray())
+          }
+          doneCount++
+        })
+      })
+    },
+
+    internalKeys: () => {
+      return internalPins.toArray()
+    },
+
+    internalKeyStrings: () => {
+      return internalPins.toStringArray()
+    },
+
+    // encodes and writes pin key sets to the datastore
+    flush: promisify((callback) => {
+      // callback (err, root)
+      const newInternalPins = new KeySet()
+      const logInternalKey = (mh) => newInternalPins.add(mh)
+      // each key set will be stored as a DAG node, and root will link to both
+      pin.set.storeSet(pin.directKeys(), logInternalKey, (err, dRoot) => {
+        if (err) { return callback(err) }
+        const dLink = new DAGLink(pin.types.direct, dRoot.size, dRoot.multihash)
+        pin.set.storeSet(pin.recursiveKeys(), logInternalKey, (err, rRoot) => {
+          if (err) { return callback(err) }
+          const rLink = new DAGLink(pin.types.recursive, rRoot.size, rRoot.multihash)
+          // the set nodes link to an empty node, so make sure it's added to dag
+          DAGNode.create(new Buffer(0), (err, emptyNode) => {
+            if (err) { return callback(err) }
+            dag.put(emptyNode, {cid: new CID(emptyNode.multihash)}, (err, cid) => {
+              if (err) { return callback(err) }
+              // create root node pointing to direct and recursive nodes
+              DAGNode.create(new Buffer(0), [dLink, rLink], (err, root) => {
+                if (err) { return callback(err) }
+                // then add the root node to dag
+                dag.put(root, {cid: new CID(root.multihash)}, (err, cid) => {
+                  if (err) { return callback(err) }
+                  // update the internal pin set
+                  logInternalKey(root.multihash)
+                  // save serialized root to datastore under a consistent key
+                  if (!repo.closed) {
+                    repo.datastore.put(pinDataStoreKey, root.serialized, (err) => {
+                      if (err) { return callback (err) }
+                      internalPins = newInternalPins
+                      return callback(null, root)
+                    })
+                  }
+                  else {
+                    // skip for now
+                    console.log('repo was closed within add callbacks')
+                    return callback(null, root)
+                  }
+                })
+              })
+            })
+          })
+        })
+      })
```

Comment:
Please use async.series or async.waterfall for this code

https://caolan.github.io/async/docs.html#